runtime/src/main/java/org/corfudb/runtime/collections/TxnContext.java

```diff
@@ -0,0 +1,435 @@
+     * @param indexName Index name. In case of protobuf-defined secondary index it is the field name.
+     * @param indexKey  Key to query.
+     * @param <K>       Type of Key.
+     * @param <V>       Type of Value.
+     * @param <I>       Type of index/secondary key.
+     * @return Result of the query.
+     */
+    @Nonnull
+    public <K extends Message, V extends Message, M extends Message, I extends Comparable<I>>
+    List<CorfuStoreEntry<K, V, M>> getByIndex(@Nonnull Table<K, V, M> table,
+                                              @Nonnull final String indexName,
+                                              @Nonnull final I indexKey) {
+        applyWritesForReadOnTable(table);
+        table.getMetrics().incNumGetByIndexes();
+        return table.getByIndex(indexName, indexKey);
+    }
+
+    /**
+     * Gets the count of records in the table at a particular timestamp.
+     *
+     * @param table - the table whose count is requested.
+     * @return Count of records.
+     */
+    public <K extends Message, V extends Message, M extends Message>
+    int count(@Nonnull final Table<K, V, M> table) {
+        applyWritesForReadOnTable(table);
+        table.getMetrics().incNumCounts();
+        return table.count();
+    }
+
+    /**
+     * Gets all the keys of a table.
+     *
+     * @param table - the table whose keys are requested.
+     * @return keyset of the table
+     */
+    public <K extends Message, V extends Message, M extends Message>
+    Set<K> keySet(@Nonnull final Table<K, V, M> table) {
+        applyWritesForReadOnTable(table);
+        table.getMetrics().incNumKeySets();
+        return table.keySet();
+    }
+
+    /**
+     * Scan and filter by entry.
+     *
+     * @param table Table< K, V, M > object on which the scan must be done.
+     * @param entryPredicate Predicate to filter the entries.
+     * @return Collection of filtered entries.
+     */
+    public <K extends Message, V extends Message, M extends Message>
+    List<CorfuStoreEntry<K, V, M>> executeQuery(@Nonnull final Table<K, V, M> table,
+                                                @Nonnull final Predicate<CorfuStoreEntry<K, V, M>> entryPredicate) {
+        applyWritesForReadOnTable(table);
+        table.getMetrics().incNumScans();
+        return table.scanAndFilterByEntry(entryPredicate);
+    }
+
+    /**
+     * Apply all pending writes (if any) to serve read queries.
+     */
+    private <K extends Message, V extends Message, M extends Message>
+    void applyWritesForReadOnTable(Table<K, V, M> tableBeingRead) {
+        tablesInTxn.add(tableBeingRead);
+        txnType |= READ_ONLY;
+        if (!operations.isEmpty()) {
+            operations.forEach(Runnable::run);
+            operations.clear();
+        }
+    }
+
+    /**
+     * Start the actual corfu transaction. Ensure there isn't one already in the same thread.
+     */
+    private void txBeginInternal() {
+        if (TransactionalContext.isInTransaction()) {
```

Comment:
is it better to ignore and not start a transaction instead of throwing an exception?  there can be cases where the same method in client code needs to start a transaction in some cases but is being called from a parent transaction in others

Reply:
then there is no way to validate if this is by intent or a bug. we want to only have one way of doing things that is to pass the TxnContext around to all those who need to share the same atomic operation.